dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java

```diff
@@ -0,0 +1,459 @@
+            moduleName = moduleName.trim();
+        }
+        return Optional.of(new ModuleName(appName, moduleName));
+    }
+
+    @Override
+    public int getConfigListenerPriority() {
+        return 1;
+    }
+
+    private ConfigValues getConfigValues() {
+        synchronized (this) {
+            if (configValues == null) {
+                configValues = new ConfigValues();
+            }
+            return configValues;
+        }
+    }
+
+    private class ConfigValues {
+        protected boolean isAll = false;
+        protected boolean isFirst = false;
+        protected boolean serverxmlmode = false;
+        protected List<ModuleName> included;
+        protected List<ModuleName> excluded;
+
+        /**
+         * Configures a {@code ModuleSelectionConfig}
+         * <p>
+         * This will read and parse the {@value Constants#MERGE_INCLUDE_CONFIG} and {@value Constants#MERGE_EXCLUDE_CONFIG} config properties from mpConfig.
+         * <p>
+         * TODO doc what it reads from server.xml
+         * <p>
+         * If the config is invalid, this method will output warning messages but still return a usable result object.
+         *
+         * @return the module selection config
+         */
+        protected ConfigValues() {
+
+            Config configFromMPConfig = ConfigProvider.getConfig(ApplicationRegistryImpl.class.getClassLoader());
+
+            OpenAPIServerXMLConfig configFromServerXML = null;
+            if (!ProductInfo.getBetaEdition()) {
+                serverxmlmode = false;
+            } else {
+                configFromServerXML = configFromServerXMLProvider.getConfiguration();
+                serverxmlmode = configFromServerXML.wasAnyConfigFound();
+            }
+
+            if (serverxmlmode) {
+                Tr.debug(this, tc, "Acquired includes statement from server.xml");
+
+                if (configFromServerXML.getConfigMode().isPresent()) {
+                    ConfigMode configMode = configFromServerXML.getConfigMode().get();
+                    if (configMode == OpenAPIServerXMLConfig.ConfigMode.All) {
+                        isAll = true;
+                    } else if (configMode == OpenAPIServerXMLConfig.ConfigMode.First) {
+                        isFirst = true;
+                    } else if (configMode == OpenAPIServerXMLConfig.ConfigMode.None) {
+                        included = Collections.emptyList();
+                    }
+                } else if (configFromServerXML.getIncludedAppsAndModules().isPresent()) {
+                    List<String> rawNames = configFromServerXML.getIncludedAppsAndModules().get();
+                    included = rawNames.stream().map((String rawName) -> parseModuleName(rawName, Constants.MERGE_INCLUDE_CONFIG))
+                                       .filter(Optional::isPresent)
+                                       .map(Optional::get)
+                                       .collect(Collectors.toList());
+                }
```

Comment:
It's possible to make it through here with `isAll` and `isFirst` being `false` and `included` being `null`.